File: poc/uvr_coreml/debug_model_output.py
#!/usr/bin/env python3
"""
ONNX/CoreMLモデルの出力を比較するため、入力と出力をダンプ
"""
import numpy as np
import librosa
import onnxruntime as ort

print("=" * 80)
print("🔍 ONNX Model Output Debug")
print("=" * 80)

# 簡単なテスト信号でモデル出力を確認
sr = 44100
duration = 1.0
t = np.linspace(0, duration, int(sr * duration))
test_signal = np.sin(2 * np.pi * 440 * t)

print(f"\n📊 テスト信号 (440Hz, 1秒):")
print(f"  Max: {np.abs(test_signal).max():.6f}")
print(f"  RMS: {np.sqrt(np.mean(test_signal**2)):.6f}")

# ステレオに変換
test_stereo = np.stack([test_signal, test_signal])

# STFT
n_fft = 4096
hop_length = 1024

print(f"\n🔄 STFT実行:")
left_stft = librosa.stft(test_stereo[0], n_fft=n_fft, hop_length=hop_length)
print(f"  STFT形状: {left_stft.shape}")

# 実部・虚部に分解
real_part = np.real(left_stft).astype(np.float32)
imag_part = np.imag(left_stft).astype(np.float32)

print(f"  Real range: {real_part.min():.6f} ~ {real_part.max():.6f}")
print(f"  Imag range: {imag_part.min():.6f} ~ {imag_part.max():.6f}")

# 2048ビンに制限 (モデル要求)
real_part = real_part[:2048, :]
imag_part = imag_part[:2048, :]

print(f"\n📊 モデル入力準備 (2048 bins):")
print(f"  Real shape: {real_part.shape}")
print(f"  Imag shape: {imag_part.shape}")

# 最初のチャンクを抽出 (256フレーム)
chunk_size = 256
if real_part.shape[1] < chunk_size:
    pad_width = chunk_size - real_part.shape[1]
    real_chunk = np.pad(real_part, ((0, 0), (0, pad_width)))
    imag_chunk = np.pad(imag_part, ((0, 0), (0, pad_width)))
else:
    real_chunk = real_part[:, :chunk_size]
    imag_chunk = imag_part[:, :chunk_size]

print(f"\n📦 チャンク (256フレーム):")
print(f"  Real chunk shape: {real_chunk.shape}")
print(f"  Imag chunk shape: {imag_chunk.shape}")

# モデル入力形式に変換 [1, 4, 2048, 256]
# チャンネル順: [Left Real, Left Imag, Right Real, Right Imag]
input_data = np.stack([
    real_chunk,  # Left Real (ch 0)
    imag_chunk,  # Left Imag (ch 1)
    real_chunk,  # Right Real (ch 2) - 同じデータ
    imag_chunk   # Right Imag (ch 3) - 同じデータ
], axis=0)

input_data = np.expand_dims(input_data, axis=0)

print(f"\n📥 ONNX入力:")
print(f"  形状: {input_data.shape}")
print(f"  データ型: {input_data.dtype}")
print(f"  値の範囲: {input_data.min():.6f} ~ {input_data.max():.6f}")

# ONNX推論
model_path = "models/onnx/UVR-MDX-NET-Inst_Main.onnx"
session = ort.InferenceSession(str(model_path))
input_name = session.get_inputs()[0].name
output_name = session.get_outputs()[0].name

print(f"\n🤖 ONNX推論実行:")
print(f"  Input name: {input_name}")
print(f"  Output name: {output_name}")

outputs = session.run([output_name], {input_name: input_data})
output_array = outputs[0]

print(f"\n📤 ONNX出力:")
print(f"  形状: {output_array.shape}")
print(f"  データ型: {output_array.dtype}")
print(f"  値の範囲: {output_array.min():.6f} ~ {output_array.max():.6f}")

# チャンネル別統計
print(f"\n📊 チャンネル別出力統計:")
for ch in range(4):
    ch_data = output_array[0, ch, :, :]
    print(f"  Ch{ch}: min={ch_data.min():.6f}, max={ch_data.max():.6f}, mean={ch_data.mean():.6f}")

# 440Hz binの出力を確認
freq_440_bin = int(440 * n_fft / sr)
print(f"\n🎵 440Hz bin [{freq_440_bin}] の出力 (最初の3フレーム):")
for frame_idx in range(3):
    print(f"  Frame {frame_idx}:")
    for ch in range(4):
        val = output_array[0, ch, freq_440_bin, frame_idx]
        print(f"    Ch{ch}: {val:.6f}")

# デバッグデータ保存
np.save('tests/python_output/model_input.npy', input_data)
np.save('tests/python_output/model_output.npy', output_array)

# テキスト形式でも保存 (最初のフレーム、ch0のみ)
np.savetxt('tests/python_output/model_input_ch0_frame0.txt', input_data[0, 0, :, 0])
np.savetxt('tests/python_output/model_output_ch0_frame0.txt', output_array[0, 0, :, 0])

print(f"\n💾 デバッグデータ保存完了:")
print(f"  tests/python_output/model_input.npy")
print(f"  tests/python_output/model_output.npy")
print(f"  tests/python_output/model_input_ch0_frame0.txt")
print(f"  tests/python_output/model_output_ch0_frame0.txt")

print("\n" + "=" * 80)

File: poc/uvr_coreml/ios_test/Sources/DebugModelOutput.swift
import Foundation
import CoreML
import Accelerate

@available(iOS 17.0, macOS 14.0, *)
func debugModelOutput() {
    print(String(repeating: "=", count: 80))
    print("🔍 CoreML Model Output Debug")
    print(String(repeating: "=", count: 80))
    
    // テスト信号生成 (440Hz, 1秒) - Pythonと同じ
    let sr: Float = 44100.0
    let duration: Float = 1.0
    let freq: Float = 440.0
    let sampleCount = Int(sr * duration)
    
    let testSignal: [Float] = (0..<sampleCount).map { i in
        let t = Float(i) / sr
        return sin(2.0 * Float.pi * freq * t)
    }
    
    print("\n📊 テスト信号 (440Hz, 1秒):")
    print("  Max: \(testSignal.map { abs($0) }.max()!)")
    
    // STFT
    let stftProcessor = STFTProcessor(fftSize: 4096, hopSize: 1024, windowType: .hann)
    let stftResult = stftProcessor.computeSTFT(audio: testSignal)
    
    print("\n🔄 STFT実行:")
    print("  形状: [\(stftResult.frequencyBins), \(stftResult.timeFrames)]")
    
    // 実部・虚部の範囲確認 (magnitude/phaseから復元)
    var realVals: [Float] = []
    var imagVals: [Float] = []
    
    for f in 0..<min(2048, stftResult.frequencyBins) {
        for t in 0..<stftResult.timeFrames {
            let mag = stftResult.magnitude[f][t]
            let ph = stftResult.phase[f][t]
            realVals.append(mag * cos(ph))
            imagVals.append(mag * sin(ph))
        }
    }
    
    print("  Real range: \(realVals.min()!) ~ \(realVals.max()!)")
    print("  Imag range: \(imagVals.min()!) ~ \(imagVals.max()!)")
    
    // モデル入力作成 (最初のチャンク)
    let chunkSize = 256
    let freqBins = 2048
    
    let inputArray = try! MLMultiArray(
        shape: [1, 4, freqBins, chunkSize] as [NSNumber],
        dataType: .float32
    )
    
    // データコピー
    for t in 0..<min(chunkSize, stftResult.timeFrames) {
        for f in 0..<freqBins {
            let mag = stftResult.magnitude[f][t]
            let ph = stftResult.phase[f][t]
            let real = mag * cos(ph)
            let imag = mag * sin(ph)
            
            inputArray[[0, 0, f, t] as [NSNumber]] = NSNumber(value: real)  // Left Real
            inputArray[[0, 1, f, t] as [NSNumber]] = NSNumber(value: imag)  // Left Imag
            inputArray[[0, 2, f, t] as [NSNumber]] = NSNumber(value: real)  // Right Real
            inputArray[[0, 3, f, t] as [NSNumber]] = NSNumber(value: imag)  // Right Imag
        }
    }
    
    // パディング
    if stftResult.timeFrames < chunkSize {
        for t in stftResult.timeFrames..<chunkSize {
            for f in 0..<freqBins {
                for ch in 0..<4 {
                    inputArray[[0, ch, f, t] as [NSNumber]] = 0
                }
            }
        }
    }
    
    print("\n📥 CoreML入力:")
    print("  形状: [1, 4, \(freqBins), \(chunkSize)]")
    
    // 入力の値範囲を確認
    var inputMin: Float = Float.infinity
    var inputMax: Float = -Float.infinity
    
    for idx in 0..<inputArray.count {
        let val = inputArray[idx].floatValue
        inputMin = min(inputMin, val)
        inputMax = max(inputMax, val)
    }
    
    print("  値の範囲: \(inputMin) ~ \(inputMax)")
    
    // CoreML推論
    let modelPath = "/Users/asatokazu/Documents/dev/mine/music/vocal_mastery_lab/poc/uvr_coreml/models/coreml/UVR-MDX-NET-Inst_Main.mlpackage"
    
    print("\n🤖 CoreML推論実行:")
    
    guard let compiledURL = try? MLModel.compileModel(at: URL(fileURLWithPath: modelPath)) else {
        print("❌ Model compile failed")
        return
    }
    
    let config = MLModelConfiguration()
    config.computeUnits = .all
    
    guard let model = try? MLModel(contentsOf: compiledURL, configuration: config) else {
        print("❌ Model load failed")
        return
    }
    
    let inputProvider = try! MLDictionaryFeatureProvider(dictionary: [
        "input_1": MLFeatureValue(multiArray: inputArray)
    ])
    
    guard let output = try? model.prediction(from: inputProvider) else {
        print("❌ Prediction failed")
        return
    }
    
    guard let outputArray = output.featureValue(for: "var_992")?.multiArrayValue else {
        print("❌ Output extraction failed")
        return
    }
    
    print("\n📤 CoreML出力:")
    print("  形状: \(outputArray.shape)")
    
    // 出力の値範囲
    var outputMin: Float = Float.infinity
    var outputMax: Float = -Float.infinity
    
    for idx in 0..<outputArray.count {
        let val = outputArray[idx].floatValue
        outputMin = min(outputMin, val)
        outputMax = max(outputMax, val)
    }
    
    print("  値の範囲: \(outputMin) ~ \(outputMax)")
    
    // チャンネル別統計
    print("\n📊 チャンネル別出力統計:")
    for ch in 0..<4 {
        var chVals: [Float] = []
        for f in 0..<2048 {
            for t in 0..<256 {
                let val = outputArray[[0, ch, f, t] as [NSNumber]].floatValue
                chVals.append(val)
            }
        }
        let chMin = chVals.min()!
        let chMax = chVals.max()!
        let chMean = chVals.reduce(0, +) / Float(chVals.count)
        print("  Ch\(ch): min=\(chMin), max=\(chMax), mean=\(chMean)")
    }
    
    // 440Hz binの出力
    let freq440Bin = 40
    print("\n🎵 440Hz bin [\(freq440Bin)] の出力 (最初の3フレーム):")
    for frameIdx in 0..<3 {
        print("  Frame \(frameIdx):")
        for ch in 0..<4 {
            let val = outputArray[[0, ch, freq440Bin, frameIdx] as [NSNumber]].floatValue
            print("    Ch\(ch): \(val)")
        }
    }
    
    // デバッグデータ保存
    let outputDir = "/Users/asatokazu/Documents/dev/mine/music/vocal_mastery_lab/poc/uvr_coreml/tests/swift_output"
    
    // Ch0, Frame0を保存
    var ch0Frame0: [Float] = []
    for f in 0..<2048 {
        let val = outputArray[[0, 0, f, 0] as [NSNumber]].floatValue
        ch0Frame0.append(val)
    }
    
    let ch0Frame0Text = ch0Frame0.map { "\($0)" }.joined(separator: "\n")
    try? ch0Frame0Text.write(toFile: "\(outputDir)/model_output_ch0_frame0_swift.txt", atomically: true, encoding: .utf8)
    
    print("\n💾 デバッグデータ保存完了:")
    print("  \(outputDir)/model_output_ch0_frame0_swift.txt")
    
    print("\n" + String(repeating: "=", count: 80))
}
